api/service.go

```diff
@@ -652,47 +825,73 @@ func (s *service) GetFile(ctx context.Context, req *pb.GetFileRequest) (*pb.GetF
 	if !ok {
 		log.Fatal("scope required")
 	}
-	file, err := s.getFileWithScope(ctx, req.Path, scope)
+	folderName, fileName := parsePath(req.Path)
+	folder, err := s.getFolderWithScope(ctx, folderName, scope)
+	if err != nil {
+		return nil, err
+	}
+	pth := path.New(filepath.Join(folder.Path, fileName))
+	if err = pth.IsValid(); err != nil {
+		return nil, err
+	}
+
+	node, err := s.ipfsClient.Unixfs().Get(ctx, pth)
 	if err != nil {
 		return nil, err
 	}
+	defer node.Close()
 
-	return fileToPbFile(file), nil
-}
+	size, err := node.Size()
+	if err != nil {
+		return nil, err
+	}
 
-func fileToPbFile(file *c.File) *pb.GetFileReply {
 	return &pb.GetFileReply{
-		ID:        file.ID,
-		Path:      file.Path,
-		Name:      file.Name,
-		ProjectID: file.ProjectID,
-		Created:   file.Created,
-	}
+		Path: pth.String(),
+		Size: size,
+	}, nil
 }
 
-// ListFiles handles a list files request.
-func (s *service) ListFiles(ctx context.Context, req *pb.ListFilesRequest) (*pb.ListFilesReply, error) {
-	log.Debugf("received list files request")
+// CatFile handles a cat file request.
+func (s *service) CatFile(req *pb.CatFileRequest, server pb.API_CatFileServer) error {
+	log.Debugf("received cat file request")
 
-	scope, ok := ctx.Value(reqKey("scope")).(string)
+	scope, ok := server.Context().Value(reqKey("scope")).(string)
 	if !ok {
 		log.Fatal("scope required")
 	}
-	proj, err := s.getProjectForScope(ctx, req.ProjectID, scope)
+	folderName, fileName := parsePath(req.Path)
+	folder, err := s.getFolderWithScope(server.Context(), folderName, scope)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	pth := path.New(filepath.Join(folder.Path, fileName))
+	if err = pth.IsValid(); err != nil {
+		return err
 	}
 
-	files, err := s.collections.Files.List(ctx, proj.ID)
+	node, err := s.ipfsClient.Unixfs().Get(server.Context(), pth)
 	if err != nil {
-		return nil, err
-	}
-	list := make([]*pb.GetFileReply, len(files))
-	for i, file := range files {
-		list[i] = fileToPbFile(file)
+		return err
 	}
+	defer node.Close()
 
-	return &pb.ListFilesReply{List: list}, nil
+	file := ipfsfiles.ToFile(node)
```

Comment:
Oh, this doesn't return an error if this path happens to be a directory?
Or maybe fails earlier?

Reply:
Oops! Yep, file will be nil if this is a directory (this method is just a helper that does an assertion). I added a `nil` check.